extensions/basic_auth/plugin.cc

```diff
@@ -0,0 +1,287 @@
+            } else if (match == "suffix") {
+              rules.pattern = Suffix;
+            }
+            rules.request_path = request_path;
+            for (auto& methods : request_methods) {
+              basic_auth_configuration_[methods].push_back(rules);
+            }
+            return true;
+          })) {
+    LOG_WARN("Failed to parse configuration information.");
+  }
+  return true;
+}
+
+FilterHeadersStatus PluginContext::credentialsCheck(
+    const PluginRootContext::BasicAuthConfigRule& rule,
+    const std::string& authorization_header) {
+  auto credential_iter = rule.encoded_credentials.find("");
+  auto credential_size = rule.encoded_credentials.size();
+  // Check if credential set is of of size 1 and its an empty string.
+  if (credential_size == 1 &&
+      credential_iter != rule.encoded_credentials.end()) {
+    deniedNoAccessToCredentials();
+    return FilterHeadersStatus::StopIteration;
+  }
+  // Check if there is an authorization header
+  if (authorization_header.compare("") == 0) {
+    deniedNoAuthorizationHeader();
+    return FilterHeadersStatus::StopIteration;
+  }
+  absl::string_view basic_authorization_header = authorization_header;
+  // Check if the Basic auth header starts with "Basic "
+  if (!absl::ConsumePrefix(&basic_authorization_header, "Basic ")) {
+    deniedNoBasicAuthData();
+    return FilterHeadersStatus::StopIteration;
+  }
+  auto auth_credential_iter =
+      rule.encoded_credentials.find(std::string(basic_authorization_header));
+  // Check if encoded credential is part of the encoded_credentials
+  // set from our container to grant or deny access.
+  if (auth_credential_iter == rule.encoded_credentials.end()) {
+    deniedInvalidCredentials();
+    return FilterHeadersStatus::StopIteration;
+  } else {
+    return FilterHeadersStatus::Continue;
+  }
+  return FilterHeadersStatus::StopIteration;
+}
+
+FilterHeadersStatus PluginContext::onRequestHeaders(uint32_t, bool) {
+  auto basic_auth_configuration = rootContext()->basicAuthConfigurationValue();
+  auto request_path = getRequestHeader(":path")->toString();
+  auto method = getRequestHeader(":method")->toString();
+  auto method_iter = basic_auth_configuration.find(method);
+  // First we check if the request method is present in our container
+  if (method_iter != basic_auth_configuration.end()) {
+    // We iterate through our vector of struct in order to find if the
+    // request_path according to given match patterns, is part of the plugin's
+    // configuration data. If that's the case we check the credentials
+    for (auto& rules : basic_auth_configuration[method]) {
+      if (rules.pattern == PluginRootContext::MATCH_TYPE::Prefix) {
+        absl::string_view path = request_path;
+        if (absl::ConsumePrefix(&path, rules.request_path)) {
+          auto authorization_header =
+              getRequestHeader("authorization")->toString();
+          return credentialsCheck(rules, authorization_header);
```

Comment:
You probably need to go through all the rules, instead of just find one match and claim done. Consider case where several rules match the same path, this needs to check the authorization header against all of them.